asllib/doc/ASLTypingReference.tex

```diff
@@ -30,50 +75,78 @@ \chapter{Copyright and Disclaimer}
+
+The purpose of this document is to describe, in a definitive way, which ASL programs are considered \emph{valid}. Validity is defined in terms of a \emph{type system}~\cite{TypeSystemsLucaCardelli}. That is, a set of \emph{type rules}.
 
 \section{Abstract Syntax}
-An \emph{abstract syntax tree} (AST, for short) represents an ASL program as a labelled tree. 
-The ASL Abstract Syntax is given in~\cite{ASLAbstractSyntaxReference}.
+An \emph{abstract syntax tree} (AST, for short) represents an ASL program as a labelled structured tree. 
+%
+The ASL type system is defined over the ASL abstract syntax, which is given in~\cite{ASLAbstractSyntaxReference}.
 
 \section{Environments}
 
-An environment is what the typing operates over: a structure which amongst
-other things associates types to variables. Intuitively, the typing of a
+An \emph{environment} is what the type rules operate over: a structure, which among
+other things, associates types to variables. Intuitively, the typing of a
 program makes an initial environment evolve, with new types as given by the
-variable declarations of the program. Environments map names to variables and
-subprograms.
+variable declarations of the program.
+
+Environments $E \in \mathbb{E}$ are formally defined below (referring to non-terminals in the abstract syntax):
+\[
+\begin{array}{rcl}
+\mathbb{E} 	&=& \mathbb{G} \times \mathbb{L}\\
+\mathbb{G} 	&=& \declaredtypes \times \constantsvalues \times \storagetypes\\
+  			& & \times\ \subtypes \times \subprograms \times \subprogramrenamings\\
+\mathbb{L} 	&=& \constantsvalues \times \storagetypes \times \returntype\\
+\hline
+\declaredtypes	 &=& \identifier \partialto \ty\\
+\constantsvalues &=& \identifier \partialto \literal\\
+\storagetypes    &=& \identifier \partialto \ty \times \localdeclkeyword\\
+\subtypes		 &=& \identifier \partialto \identifier\\
+\subprograms	 &=& \identifier \partialto \func\\
+\subprogramrenamings	 &=& \identifier \rightarrow 2^\func\\
+\returntype      &=& \{\langle\rangle, \langle \ty \rangle\}
+\end{array}
+\]
+
+Technically, an environment $E=(G^E, L^E)$ consists of two
+distinct components: the global environment $G^E$---pertaining to AST nodes
+appearing outside of a given subprogram, and the local environment
+$L^E$---pertaining to AST nodes appearing inside a given subprogram.
+%
+We will drop the superscripts when they are obvious from the context.
+%
+The global environment and local environment consist of various maps. We will use the notation $G.m$ and $L.m$ to access the $m$ component of a given environment.
 
-More precisely, a \emph{typing environment} (environment, for short) maps AST
-nodes to types.
 
 A \emph{type judgement} associates a node in an abstract syntax tree (which
-intuitively corresponds to an ASL construct) with a certain type. We then say
-that the node is \emph{annotated} with that type.
+intuitively corresponds to an ASL construct) with a \emph{sentence}. That is, a statement that holds true.
```

Comment:
please consider rephrasing this - perhaps "verdict"?
also please expand: "A verdict in this context is a statement that holds true."